File: SIUCF24TA/verifying/polyominoes/polyominoes.cpp
//Solution for polyominoes
//Thomas Meeks


//This is very much an implementation problem and i am not sorry for making it.
//The main idea is to build each Degree N+1 polyomino from the preceeding
//Degree N polyominoes. To avoid double counting we need to canonicalize
//that means for any given polyomino that is the same as another we need to
//transform them both into the same representitive polyomino
//since in this solution the polyomonioes are represented by a unsigned long long
//and are on an 8 by 8 grid a form of canonicalization we can use
//is all reflections and rotations, shift the shape to an arbitary corner
//and then for all of those resulting masks we will use the mask with the
//minimum numeric value, minimum is arbitary it could also be maximum
//we will need alot of helper functions to handle all of the bitwise operations needed for moving the mask

//if you don't fully understand bitwise operations i highly reccomend learning them before trying
//to decipher this code.
#include <bits/stdc++.h>
#include <iostream>
using namespace std;
//the 64 bit unsigned long long represented an 8by8 grid in row major order
typedef unsigned long long ull;

//get a particular cell in the grid (from it's mask form)
inline bool get(int r, int c, ull &mask){
    return mask & (1ul << ((r*8)+c));
}

//set a particular cell in the grid (from it's mask form)
inline void setM(int r, int c, ull &mask){
    mask |= (1ul << ((r*8)+c));
}

//clear a particular cell in the grid (from it's mask form)
inline void clearM(int r, int c, ull &mask){
    mask &= ~(1ul << ((r*8)+c));
}

//helper function for displaying a grid, only used for debugging
void display(ull mask){
    for(int r = 0; r < 8; r++){
        for(int c = 0; c < 8; c++){
            if(get(r,c,mask)) cout << "X";
            else cout << "O";
        }
        cout << "\n";
    }
}
//a whole lot of global variables 
ull maskExpand[256];
ull mask8Reflect[256];
const int sizeSquare = 1<<16;
ull squarePreComp[sizeSquare];
ull borderRing = 0;
ull lastBitMask = 0;
ull firstBitMask = 0;
ull leftBorder = 0;
ull rightBorder = 0;
ull topBorder = 0;
ull bottomBorder = 0;

//allows us to do rotations in "8" operations rather than 64, the rotation ends up being
//taking each group of 8 bits and expanding them out to be a fully mask, each group has a different offset
//we have an array of precomped bit expansions.
void rotate(ull &mask){
    ull newMask = 0;
    ull eightMask = (1<<8)-1;
    for(int groupIndex = 0; groupIndex < 8; groupIndex++){
        ull groupMask = ((mask >> (groupIndex * 8)) & eightMask);
        newMask |= (maskExpand[groupMask] << groupIndex);
    }
    mask = newMask;
}
//implace and function versions of shift up and shiftdown because i could't decide which one i was using
void shiftDown(ull &mask, int rows = 1){ mask = (mask << (8*rows));  }
void shiftUp(ull &mask, int rows = 1) {mask = (mask >> (8*rows)); }
ull shiftedUp(ull mask, int rows = 1) {mask = (mask >> (8*rows)); return mask; }
ull shiftedDown(ull mask, int rows = 1){ mask = (mask << (8*rows)); return mask;  }
//when shifting right or left we first clear the bits at the end so they don't move to the next row.
void shiftRight(ull &mask, int cols = 1){
    mask &= ~(lastBitMask);
    mask <<= (cols);
}

void shiftLeft(ull &mask, int cols = 1){
    mask &= ~(firstBitMask);
    mask >>= (cols);
}

ull shiftedRight(ull mask, int cols = 1){
    mask &= ~(lastBitMask);
    mask <<= (cols);
    return mask;
}

ull shiftedLeft(ull mask, int cols = 1){
    mask &= ~(firstBitMask);
    mask >>= (cols);
    return mask;
}

//flipping each group of 8
void reflectVerticle(ull &mask){
    ull newMask = 0;
    ull mask8 = (1<<8)-1;
    newMask |= (((mask8) & mask) << 56);
    newMask |= (((mask8<<8) & mask) << 40);
    newMask |= (((mask8<<16) & mask) << 24);
    newMask |= (((mask8<<24) & mask) << 8);

    newMask |= (((mask8<<56) & mask) >> 56);
    newMask |= (((mask8<<48) & mask) >> 40);
    newMask |= (((mask8<<40) & mask) >> 24);
    newMask |= (((mask8<<32) & mask) >> 8);
    mask = newMask;
}
//i thought i needed both reflections
//you actually only need one reflection so this is useless
void reflectHorizontal(ull &mask){
    ull newMask = 0;
    ull mask8 = (1<<8)-1;
    for(int i = 0; i < 8; i++){
        newMask |= mask8Reflect[(mask>>(i*8))&mask8] << (i*8);
    }
    mask = newMask;
}

void precomp(){
    //memset our arrays to zero because it helps me sleep better at night
    memset(maskExpand,0,sizeof(maskExpand));
    memset(mask8Reflect,0,sizeof(mask8Reflect));
    memset(squarePreComp,0,sizeof(squarePreComp));

    //make mask for the first and last bit of each row for left and right shift clearing
    for(int i = 0; i < 8; i++) firstBitMask |= (1ul << (i*8));
    for(int i = 0; i < 8; i++) lastBitMask |= (1ul << ((i*8)|7));
    
    //i promise this is a square <3
    ull square = 3|(3<<8);
    //ok basically instead of looking through all spots in the grid looking for a square we can
    //precomp 2 rows (2^16 = 65536) and now we only need to check 7 array slots when looking for a square
    for(int mask = 0; mask < (1<<16); mask++){
        for(int i = 0; i < 7; i++){
            if(((square << i) & mask) == (square << i)) squarePreComp[mask] = true; 
        }
    }

    //lord forgive me
    for(int mask = 0; mask < (1<<8); mask++){
        //not the shortest way of doing this but copy and paste go brrr
        ull bit0 = (bool)(mask & (1<<0));
        ull bit1 = (bool)(mask & (1<<1));
        ull bit2 = (bool)(mask & (1<<2));
        ull bit3 = (bool)(mask & (1<<3));
        ull bit4 = (bool)(mask & (1<<4));
        ull bit5 = (bool)(mask & (1<<5));
        ull bit6 = (bool)(mask & (1<<6));
        ull bit7 = (bool)(mask & (1<<7));
        //useless reflection mask that i didn't even need
        mask8Reflect[mask] =(bit7 << 0)  | (bit6 << 1)  |
                            (bit5 << 2)  | (bit4 << 3)  |
                            (bit3 << 4)  | (bit2 << 5)  |
                            (bit1 << 6)  | (bit0 << 7);
        //expand out the mask for the rotation earlier in the code
        maskExpand[mask] =  (bit7 << 0)  | (bit6 << 8)  |
                            (bit5 << 16) | (bit4 << 24) |
                            (bit3 << 32) | (bit2 << 40) |
                            (bit1 << 48) | (bit0 << 56);
    }
    //border ring ends up not being needed
    for(int c = 0; c < 8; c++){
        setM(0,c,borderRing);
        setM(7,c,borderRing);
        setM(0,c,topBorder);
        setM(7,c,bottomBorder);
    }

    for(int r = 0; r < 8; r++){
        setM(r,0,borderRing);
        setM(r,7,borderRing);
        setM(r,0,leftBorder);
        setM(r,7,rightBorder);
    }
}


ull moveToCorner(ull mask){
    //we are gonna keep moving the mask to the left and down until it hits the border
    //once it gets to the border we will push it back up so it's not on the edge and
    //new cells can still be placed around it
    while((mask & leftBorder) == 0){ shiftLeft(mask); }
    while((mask & bottomBorder) == 0) shiftDown(mask);
    if(mask & leftBorder) shiftRight(mask);
    if(mask & bottomBorder) shiftUp(mask);
    return mask;
}
//what a nightmare
ull minTransformation(ull mask){
    ull res = ULLONG_MAX;
    //rather clunky since i decided to make this pass by reference
    ull reflection = mask; reflectVerticle(reflection);
    mask = moveToCorner(mask);
    //always move to corner before minning result!!
    res = min(min(res,moveToCorner(mask)),moveToCorner(reflection));
    rotate(mask); res = min(res,moveToCorner(mask));
    rotate(mask); res = min(res,moveToCorner(mask));
    rotate(mask); res = min(res,moveToCorner(mask));
    rotate(reflection); res = min(res,moveToCorner(reflection));
    rotate(reflection); res = min(res,moveToCorner(reflection));
    rotate(reflection); res = min(res,moveToCorner(reflection));
    return res; //which ever result is the smallest was the canonical representation
}

bool valid(ull mask){
    mask = moveToCorner(mask);
    if((mask & leftBorder) || (mask & rightBorder)) return false;
    if((mask & topBorder) || (mask & bottomBorder)) return false;
    //return true;
    //speedy square check because of precomp
    ull mask16 = (1<<16)-1;
    for(int i = 0; i < 7; i++){
        if(squarePreComp[mask16 & (mask >> (i*8))]) return false;
    }

    return true;
}

int main(){
    #ifndef ONLINE_JUDGE
        freopen("thomas.out","w", stdout);
    #endif
    //precomppppp
    precomp();
    //tis is where all the magic happens
    int n; cin >> n;
    vector<vector<ull> > results(n+1);
    ull startingGuy = 0; 
    setM(4,4,startingGuy);
    startingGuy = minTransformation(startingGuy); //maybe uneccesary but :shrug:
    results[1].push_back(startingGuy); //the starting guy
    vector<int> ans(n+1); ans[1] = 1; 
    //now that we have all of our helper functions the code from here is pretty easy
    for(int degree = 2; degree <= n; degree++){
        for(ull poly: results[degree-1]){
            ull neighbors = 0;  //bit mask of all valid neighbors
            //well a neighbor has to be next to the current poly so lets just shake it around
            neighbors |= shiftedDown(poly);
            neighbors |= shiftedUp(poly);
            neighbors |= shiftedLeft(poly);
            neighbors |= shiftedRight(poly);
            neighbors &= ~(poly); //we can't place a square on ourselves so lets get rid of that

            //might be a better way of doing this but you can have tons of neighbors anyway
            for(int i = 0; i < 64; i++){
                if(neighbors & (1ul << i)){
                    ull newPoly = poly|(1ul<<i);
                    if(!valid(newPoly)) continue; //if the newpolygon isn't valid lets not add it to our list

                    newPoly = minTransformation(newPoly); //find canonical representation
                    results[degree].push_back(newPoly); //lets goooo
                }
            }
        }

        sort(results[degree].begin(),results[degree].end());
        //after items are sorted there is built in functions to uniqueify the vector
        results[degree].erase(unique(results[degree].begin(),results[degree].end()),results[degree].end());
        ans[degree] = results[degree].size(); //we found our answer
        if(degree-1 >= 0) results[degree-1].clear(); 
        //lets just get rid of these guys to save memory not that big of deal since the growth is exponential but
        //save the turles amirite
        //for(auto el: results[degree]) { display(el); cout << '\n'; }
        //cout << degree << " " << ans[degree] << "\n";
        //int x; cin >> x;
    }
    sort(results[n].begin(), results[n].end());
    // display(results[12][0]);
    for(int i = 0; i < results[n].size(); i++){
        cout << results[n][i] << endl;
    }
    //oh look we can now print the answer
    // for(int i = 1; i< ans.size(); i++) cout << ans[i] << "\n";
    // return 0;
}

File: SIUCF24TA/verifying/polyominoes/polyominoes.py
fi = open("tyler.in", "r")
# fo = open("brady.out", "w")

def input():
    return fi.readline()

# def print(s, end = '\n'):
#     fo.write(str(s) + end)

# Written by: Brady Aber
# UCF SI Camp 24
# 7/11/24
# polyominoes.py

mxPoly = 1<<64

# A nice way to print a polyomino
def prnt(n):
    for i in reversed(range(8)):
        for j in range(8):
            print('⬜' if n&((1<<(8*i))<<j) else '⬛', end = '')
        print("")
    print("")

# Bitwise shifts a polyomino up
def up(n, force=False):
    top = 1<<48
    if not force:
        for i in range(8):
            if n&top:
                return n
            top<<=1
    return n<<8

# Bitwise shifts a polyomino down
def down(n, force=False):
    bottom = 1<<8
    if not force:
        for i in range(8):
            if n&bottom:
                return n
            bottom<<=1
    return n >> 8

# Bitwise shifts a polyomino right
def right(n, force=False):
    side = 1<<6
    if not force:
        for i in range(8):
            if n&side:
                return n
            side<<=8
    return n<<1

# Bitwise shifts a polyomino left
def left(n, force=False):
    side = 1<<1
    if not force:
        for i in range(8):
            if n&side:
                return n
            side<<=8
    return n>>1

# Rotates a polyomino to the right
def rotateRight(n):
    newN = 0
    for i in range(8):
        for j in range(8):
            if n&1<<(8*j + 7-i):
                newN |= 1<<(8*i + j)
    return newN

# Horizontally flips a polyomino
def flipHorizontal(n):
    newN = 0
    for i in range(8):
        for j in range(8):
            if n&1<<(8*i + j):
                newN |= 1<<(8*i + 7-j)
    return newN

# Make a new polyomino in a certain direction given a function and its inverse
def createNewPolyominoes(n, f, finv):
    inv = finv(n)
    newSpaces = (f(inv, force=True)&(-inv - 1))
    newSpaces&=35604928818740736
    last = 1
    out = []
    for shift in range(64):
        if newSpaces&last:
            out.append(last|inv)
        last <<= 1
    return out

# Canonicalizes a polyomino, by moving it to the top left, rotating and flipping it
# Takes the min of all 8 of these, which will guarentee all 8 give the same number
def canonicalization(n):
    flippedN = flipHorizontal(n)
    canons = []
    canons.append(n)
    canons.append(flippedN)
    for i in range(3):
        n = rotateRight(n)
        flippedN = rotateRight(flippedN)
        canons.append(n)
        canons.append(flippedN)
    for i in range(len(canons)):
        
        leftShift = left(canons[i])
        while leftShift!=canons[i]:
            canons[i] = leftShift
            leftShift = left(leftShift)

        downShift = up(canons[i])
        while downShift!=canons[i]:
            canons[i] = downShift
            downShift = up(downShift)
    return min(canons)

# Checks for a 2x2 square throughout the polyomino
def containsSquare(n):
    square = 394752
    # print(square)
    for i in range(6):
        tempsquare = square
        for j in range(6):
            if n&tempsquare == tempsquare:
                return True
            tempsquare <<= 1
        square<<=8
    return False


polyominoes = [[], [right(up(1))]]

# n = int(input())
n = 5
# print(1)
# Compute them in order
for i in range(2, n+1):
    nextGen = set()

    # For all the polyominoes, add the canonicalization of what they can turn into
    for poly in polyominoes[i-1]:
        newAdds = []
        newAdds.extend(createNewPolyominoes(poly, up, down))
        newAdds.extend(createNewPolyominoes(poly, down, up))
        newAdds.extend(createNewPolyominoes(poly, right, left))
        newAdds.extend(createNewPolyominoes(poly, left, right))

        # Add the canonicalized versions of all polyominoes to the set 
        for poly in newAdds:
            c = canonicalization(poly)
            if not containsSquare(c):
                # The set makes sure there are no duplicate canonicalizations
                nextGen.add(c)

    polyominoes.append(list(nextGen))
    # print(len(polyominoes[-1]))
# s = set(polyominoes[11])
# tyler = set()
# for i in range(6655):
#     poly = 0
#     for j in range(8):
#         inp = input().strip()
#         for k in range(8):
#             if inp[k]=='1':
#                 poly += 1<<k
#         poly<<=8
#     poly>>=8
#     input()
#     poly = canonicalization(right(down(poly)))
#     tyler.add(poly)
#     # prnt(poly)

# for poly in s:
#     if poly not in tyler:
#         print(poly)

prnt(569566283695104)
prnt(565157802047488)
prnt(565174949408768)
prnt(578377618046976)
prnt(569600576587776)
prnt(565157600590848)